android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/BluetoothSite.java

```diff
@@ -0,0 +1,164 @@
+  }
+
+  /**
+   * Connects to the Gatt service of the device to download a web page and displays a progress bar
+   * for the title.
+   * @param deviceAddress The mac address of the bar
+   * @param title The title of the web page being downloaded
+   */
+  public void connect(String deviceAddress, String title) {
+    String progressTitle = activity.getString(R.string.page_loading_title) + title;
+    progress = ProgressDialog.show(activity, progressTitle,
+        activity.getString(R.string.page_loading_message), true);
+    BluetoothAdapter.getDefaultAdapter().getRemoteDevice(deviceAddress)
+        .connectGatt(activity, false, this);
+  }
+
+  @Override
+  public void onCharacteristicRead(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic,
+      int status) {
+    if (status == BluetoothGatt.GATT_SUCCESS
+        && characteristic.getValue().length < transferRate) {
+      // Transfer is complete
+      html.append(new String(characteristic.getValue()));
+      progress.dismiss();
+      gatt.close();
+      File websiteDir = new File(activity.getFilesDir(), "Websites");
+      websiteDir.mkdir();
+      File file = new File(websiteDir, "website.html");
+      writeToFile(file);
+      if (file != null) {
+        openInChrome(file);
+      }
+    } else if (status == BluetoothGatt.GATT_SUCCESS) {
+      // Full packet received, check for more data
+      html.append(new String(characteristic.getValue()));
+      gatt.readCharacteristic(this.characteristic);
+    } else {
+      Log.i(TAG, "onCharacteristicRead unsuccessful: " + status);
+      gatt.disconnect();
+      progress.dismiss();
+      Toast.makeText(activity, R.string.ble_download_error_message, Toast.LENGTH_SHORT).show();
+    }
+  }
+
+  @Override
+  public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
+    if (newState == BluetoothProfile.STATE_CONNECTED) {
+      Log.i(TAG, "Connected to GATT server");
+      html = new StringBuilder("");
+      if (VERSION.SDK_INT >= VERSION_CODES.LOLLIPOP) {
+        gatt.requestConnectionPriority(BluetoothGatt.CONNECTION_PRIORITY_HIGH);
+        gatt.requestMtu(505);
+      } else {
+        gatt.discoverServices();
+      }
+    } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
+      Log.i(TAG, "Disconnected to GATT server");
+    }
+  }
+
+
+  @Override
+  public void onMtuChanged(BluetoothGatt gatt, int mtu, int status) {
+    Log.i(TAG, "MTU changed to " + mtu);
+    transferRate = mtu - 5;
+    gatt.discoverServices();
+  }
+
+  @Override
+  public void onServicesDiscovered(BluetoothGatt gatt, int status) {
+    Log.i(TAG, "Services Discovered");
+    if (status == BluetoothGatt.GATT_SUCCESS) {
+      List<BluetoothGattService> services = gatt.getServices();
+      for (BluetoothGattService gattService : services) {
+        if (gattService.getUuid().toString().equals(SERVICE_UUID)) {
+          characteristic = gattService.getCharacteristics().get(0);
```

Comment:
@hayesjordan @scottjenson Could we at least define a characteristic UUID instead of simply getting the "first" one?